File: src/Systems/Zig/adapter.zig
const std = @import("std");

const LegacyFahrenheitSensor = struct {
    fn readFahrenheit(_: *const LegacyFahrenheitSensor) i32 {
        return 86;
    }
};

const FahrenheitSensorAdapter = struct {
    adaptee: *const LegacyFahrenheitSensor,

    fn readCelsius(self: *const FahrenheitSensorAdapter) i32 {
        const fahrenheit = self.adaptee.readFahrenheit();
        return @divTrunc((fahrenheit - 32) * 5, 9);
    }
};

pub fn main() void {
    const legacy = LegacyFahrenheitSensor{};
    const adapter = FahrenheitSensorAdapter{ .adaptee = &legacy };

    std.debug.print("legacy={d}F\n", .{legacy.readFahrenheit()});
    std.debug.print("adapted={d}C\n", .{adapter.readCelsius()});
}

File: src/Niche/Nim/adapter.nim
type
  LegacyFahrenheitSensor = object
  FahrenheitSensorAdapter = object
    adaptee: LegacyFahrenheitSensor

proc readFahrenheit(_: LegacyFahrenheitSensor): int =
  86

proc readCelsius(adapter: FahrenheitSensorAdapter): int =
  let fahrenheit = adapter.adaptee.readFahrenheit()
  ((fahrenheit - 32) * 5) div 9

let legacy = LegacyFahrenheitSensor()
let adapter = FahrenheitSensorAdapter(adaptee: legacy)

echo "legacy=", legacy.readFahrenheit(), "F"
echo "adapted=", adapter.readCelsius(), "C"
